modules/regexp-parser/regexp-parser.c

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2021 Balabit
+ * Copyright (c) 2021 Xiaoyu Qiu
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "regexp-parser.h"
+#include "parser/parser-expr.h"
+#include "scratch-buffers.h"
+#include "string-list.h"
+
+#include <string.h>
+#include <stdio.h>
+
+typedef struct _RegexpParser
+{
+  LogParser super;
+  gchar *prefix;
+  GList *patterns;
+  LogMatcherOptions matcher_options;
+  GList *matchers;
+} RegexpParser;
+
+LogMatcherOptions *
+regexp_parser_get_matcher_options(LogParser *s)
+{
+  RegexpParser *self = (RegexpParser *) s;
+
+  return &self->matcher_options;
+}
+
+void
+regexp_parser_set_prefix(LogParser *s, const gchar *prefix)
+{
+  RegexpParser *self = (RegexpParser *) s;
+
+  g_free(self->prefix);
+  if (prefix)
+    self->prefix = g_strdup(prefix);
+  else
+    self->prefix = NULL;
+}
+
+void
+regexp_parser_set_patterns(LogParser *s, GList *patterns)
+{
+  RegexpParser *self = (RegexpParser *) s;
+
+  string_list_free(self->patterns);
+  self->patterns = patterns;
+}
+
+gboolean
+regexp_parser_compile(LogParser *s)
+{
+  RegexpParser *self = (RegexpParser *) s;
+
+  log_matcher_options_init(&self->matcher_options);
+  for (GList *item = self->patterns; item; item = item->next)
+    {
+      self->matchers = g_list_prepend(self->matchers, log_matcher_new(&self->matcher_options));
+      log_matcher_pcre_set_prefix((LogMatcher *)self->matchers->data, self->prefix);
+
+      if(!log_matcher_compile((LogMatcher *)self->matchers->data, item->data, NULL))
```

Comment:
do we handle correctly the case where the 3rd pattern fails to compile? any memory leaks?

Reply:
@bazsi I'm a little confused about this problem. Does the memory leak here mean that the space allocated by `log_matcher_new` has not been released? But if any pattern fails to compile, the program will terminal directly and print the error on the console.

Reply:
I just wanted to confirm that if the loop exits early (because of log_matcher_compile() returning failure), no memory leaks remain.

It is true that the exiting process would free all allocations, but we should ensure that all internal objects do that on their own without relying on the process to exit.

In any case, just make sure that log_matcher_free() is properly called for the partially populated list too.